app/gui2/shared/ast/tree.ts

```diff
@@ -749,6 +749,59 @@ export interface MutableUnaryOprApp extends UnaryOprApp, MutableAst {
 }
 applyMixins(MutableUnaryOprApp, [MutableAst])
 
+interface AutoscopedIdentifierFields {
+  operator: NodeChild<SyncTokenId>
+  identifier: NodeChild<SyncTokenId>
+}
+export class AutoscopedIdentifier extends Ast {
+  declare fields: FixedMapView<AstFields & AutoscopedIdentifierFields>
+  constructor(module: Module, fields: FixedMapView<AstFields & AutoscopedIdentifierFields>) {
+    super(module, fields)
+  }
+
+  static tryParse(
+    source: string,
+    module?: MutableModule,
+  ): Owned<MutableAutoscopedIdentifier> | undefined {
+    const parsed = parse(source, module)
+    if (parsed instanceof MutableAutoscopedIdentifier) return parsed
+  }
+
+  static concrete(module: MutableModule, operator: NodeChild<Token>, identifier: NodeChild<Token>) {
+    const base = module.baseObject('AutoscopedIdentifier')
+    const fields = composeFieldData(base, {
+      operator,
+      identifier,
+    })
+    return asOwned(new MutableAutoscopedIdentifier(module, fields))
+  }
+
+  static new(module: MutableModule, operator: Token, identifier: Token) {
+    return this.concrete(module, unspaced(operator), unspaced(identifier))
+  }
+
+  *concreteChildren(_verbatim?: boolean): IterableIterator<RawNodeChild> {
+    const { operator, identifier } = getAll(this.fields)
+    yield operator
+    if (identifier) yield identifier
+  }
+}
+export class MutableAutoscopedIdentifier extends AutoscopedIdentifier implements MutableAst {
+  declare readonly module: MutableModule
+  declare readonly fields: FixedMap<AstFields & AutoscopedIdentifierFields>
+
+  setOperator(value: Token) {
+    this.fields.set('operator', unspaced(value))
+  }
+  setIdentifier(value: Token) {
```

Comment:
It's important that we don't allow any mutations that would break the syntax--we must only accept values that, if printed and re-parsed, would still yield an `AutoscopedIdentifier`. In this case the only legal operands are capitalized `Identifier`s; a subtype of `Identifier` could be defined to enforce this restriction, maybe called `TypeOrConstructorIdentifier` (I have never been able to come up with a less awkward name for the things that are capitalized in Enso syntax).